### PR TITLE
Convert "Visualization utilities" notebook into a sphinx-gallery example

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR converts https://github.com/pytorch/vision/blob/master/examples/python/visualization_utils.ipynb into a sphinx-gallery example.

I modified it a bit to keep it short and to the point. I also chose not to showcase `make_grid` because I don't see a compelling use-case for it: users can just create matplotlib grid themselves, the mpl API has gotten a lot better in the last few years. Also it's impossible to plot images with different dimensions wit hit. (No strong opinion, feel free to disagree, I can put it back).

Eventually I think we'll want to remove the original notebook but I'm keeping it there for now.

Image used is https://www.pexels.com/photo/smiling-women-and-men-sitting-on-green-grass-1231230/

CC @oke-aditya  as I think you wrote the original notebook?

### CODE REVIEW

File: gallery/plot_visualization_utils.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+=======================
+Visualization utilities
+=======================
+
+This example illustrates some of the utilities that torchvision offers for
+visualizing bounding boxes and segmentation masks.
+"""
+
+
+from pathlib import Path
```
Comment (oke-aditya) [inline gallery/plot_visualization_utils.py:11]: I think this import is unnecessary. Haven't checked in IDE though.

**Comment by oke-aditya** [comment]
Yes @NicolasHug  I had written the notebook :smile: 

I think `make_grid` is quite old but still seen people using it, so l think we can keep it

People often ask about this in [dicusssion forums](https://discuss.pytorch.org/t/what-is-the-correct-way-of-using-trchvision-utils-make-grid/55117/4) [stackoveflow](https://stackoverflow.com/questions/51329159/how-can-i-generate-and-display-a-grid-of-images-in-pytorch-with-plt-imshow-and-t)


And even in  some [codebases](https://www.programcreek.com/python/example/92879/torchvision.utils.make_grid)

So for legacy purpose. I think we can keep it.

File: gallery/plot_visualization_utils.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+# Naturally, we can also plot bounding boxes produced by torchvision detection
+# models.  Here is demo with a Faster R-CNN model loaded from
+# :func:`~torchvision.models.detection.fasterrcnn_resnet50_fpn`
+# model. You can also try using a RetinaNet with
+# :func:`~torchvision.models.detection.retinanet_resnet50_fpn`.
+
+from torchvision.models.detection import fasterrcnn_resnet50_fpn
+
+
+model = fasterrcnn_resnet50_fpn(pretrained=True)
```
Comment (oke-aditya) [inline gallery/plot_visualization_utils.py:61]:
```suggestion
model = fasterrcnn_resnet50_fpn(pretrained=True, progress=False)
```

File: gallery/plot_visualization_utils.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+# (:func:`~torchvision.models.segmentation.lraspp_mobilenet_v3_large`).
+#
+# Like :func:`~torchvision.utils.draw_bounding_boxes`,
+# :func:`~torchvision.utils.draw_segmentation_masks` requires a single RGB image
+# of dtype `uint8`.
+
+from torchvision.models.segmentation import fcn_resnet50
+
+
+model = fcn_resnet50(pretrained=True)
```
Comment (oke-aditya) [inline gallery/plot_visualization_utils.py:109]:
```suggestion
model = fcn_resnet50(pretrained=True, progress=False)
```

File: gallery/plot_visualization_utils.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+model = model.eval()
+output = model(umbrellas.unsqueeze(0))
+masks = output['out'].squeeze(0)
+
+umbrellas_int = T.ConvertImageDtype(dtype=torch.uint8)(umbrellas)
+result = draw_segmentation_masks(umbrellas_int, masks, alpha=0.2)
+show(result)
```
Comment (oke-aditya) [inline gallery/plot_visualization_utils.py:116]: Maybe we add one more example to show how `alpha` affects the masks?.

```suggestion

show(result)
result2 = draw_segmentation_masks(umbrellas_int, masks, alpha=0.6)
show(result2)
```

**Review by oke-aditya — COMMENTED** [review]
Few thoughts from my side @NicolasHug :smile: 

Btw torchvision gallery is awesome way to add documentation.

File: gallery/plot_visualization_utils.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
+"""
+=======================
+Visualization utilities
+=======================
+
+This example illustrates some of the utilities that torchvision offers for
+visualizing bounding boxes and segmentation masks.
```
Comment (oke-aditya) [inline gallery/plot_visualization_utils.py:7]: Maybe refolmulate the line since `make_grid` is added?
(I'm not an expert in English :sweat_smile: )

File: gallery/plot_visualization_utils.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+# The boxes are in ``(xmin, ymin, xmax, ymax)`` format
+# from torchvision.utils import draw_bounding_boxes
+
+from torchvision.utils import draw_bounding_boxes
+
+
+boxes = torch.tensor([[100, 400, 500, 740], [500, 200, 800, 580]], dtype=torch.float)
+labels = ["grass", "racoon"]
+colors = ["blue", "yellow"]
+result = draw_bounding_boxes(racoon_int, boxes, labels=labels, colors=colors, width=10)
```
Comment (oke-aditya) [inline gallery/plot_visualization_utils.py:55]: We also have `fill=True`where boxes are filled. It might be reptitive but good to illustrate?

Also Note to add (if necessary)
flled images are filled by transparency and hence should be saved with `alpha` channel.

Comment (NicolasHug) [reply gallery/plot_visualization_utils.py:55]: TBH I think it's best to try to keep the example short. We already illustrate `draw_bounding_boxes` a bunch of times, and showing the effect of every single parameter has a diminished value and can lead to a "too much info" effect. Since there's an obvious link to the docstring of the function, users will be able to discover the rest of the features easily.

Comment (oke-aditya) [reply gallery/plot_visualization_utils.py:55]: Sounds fine

File: gallery/plot_visualization_utils.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+model = model.eval()
+output = model(people.unsqueeze(0))
+masks = output['out'].squeeze(0)
+
+people_int = T.ConvertImageDtype(dtype=torch.uint8)(people)
+result = draw_segmentation_masks(people_int, masks, alpha=0.2)
+show(result)
+
+#####################################
+# We can adjust alpha to show the masks with a different transparency level:
```
Comment (fmassa) [inline gallery/plot_visualization_utils.py:125]: I think another thing which would be good to illustrate in the example is how to get the output of Mask R-CNN to be plotted with `draw_segmentation_masks`. People have asked it in the past and I think would be a good thing to showcase as well, as the output format is slightly different

File: torchvision/utils.py [hunk 8]
```diff
@@ -40,10 +40,6 @@ def make_grid(
 
     Returns:
         grid (Tensor): the tensor containing grid of images.
-
-    Example:
```
Comment (fmassa) [inline torchvision/utils.py:44]: Can we link to the galeries from the docs? I think it would be a nice thing to have

File: gallery/plot_visualization_utils.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+# model. You can also try using a RetinaNet with
+# :func:`~torchvision.models.detection.retinanet_resnet50_fpn`.
+
+from torchvision.models.detection import fasterrcnn_resnet50_fpn
+
+
+model = fasterrcnn_resnet50_fpn(pretrained=True, progress=False)
+model = model.eval()
+
+outputs = model(racoon.unsqueeze(0))
```
Comment (fmassa) [inline gallery/plot_visualization_utils.py:71]: Can we use an image which contains categories present in COCO, like dog or people? The current predictions seems a bit arbitrary.

Also, I think it would be good to showcase how to add `labels` (containing for example the class name and score).

File: gallery/plot_visualization_utils.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+# Like :func:`~torchvision.utils.draw_bounding_boxes`,
+# :func:`~torchvision.utils.draw_segmentation_masks` requires a single RGB image
+# of dtype `uint8`.
+
+from torchvision.models.segmentation import fcn_resnet50
+
+
+model = fcn_resnet50(pretrained=True, progress=False)
+model = model.eval()
+output = model(people.unsqueeze(0))
```
Comment (fmassa) [inline gallery/plot_visualization_utils.py:117]: This is not fully correct I believe, as we need to normalize with the imagenet mean / std, see https://github.com/pytorch/vision/blob/d6fee5a4c4518ddb11f3dc1b52c102a7bdad5b49/references/segmentation/presets.py#L28 .

The only model that doesn't require this is Faster-RCNN / Mask-RCNN, but classification and segmentation models require it

File: gallery/plot_visualization_utils.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+masks = output['out'].squeeze(0)
+
+people_int = T.ConvertImageDtype(dtype=torch.uint8)(people)
```
Comment (fmassa) [inline gallery/plot_visualization_utils.py:120]: In general, I would maybe recommend users using the functional API in situations like this, so something like
```python
people_int = convert_image_dtype(people, dtype=torch.uint8)
```

**Review by fmassa — COMMENTED** [review]
This looks great, thanks a ton!

I have a few comments (which can be addressed in a follow-up PR), let me know what you think

File: torchvision/utils.py (hunk 8 above)
Comment (NicolasHug) [reply torchvision/utils.py:44]: We can!
We can add manual links to an example and even reference a specific plot of a given example.

But in this case, I made sure that every torchvision function and class that is used in an example will automatically generate a "minigallery"

https://github.com/pytorch/vision/blob/d6fee5a4c4518ddb11f3dc1b52c102a7bdad5b49/docs/source/conf.py#L284-L289

image

Comment (fmassa) [reply torchvision/utils.py:44]: Awesome!

File: gallery/plot_visualization_utils.py (hunk 7 above)
Comment (NicolasHug) [reply gallery/plot_visualization_utils.py:125]: I'm getting poor results with maskrcnn_resnet50_fpn, is this expected?
image

**Comment by NicolasHug** [comment]
Thanks for the reviews! I'll add instructions for the Mask-RCNN model in a follow up PR

File: gallery/plot_visualization_utils.py (hunk 7 above)
Comment (fmassa) [reply gallery/plot_visualization_utils.py]: Sure, no worries, thanks for all your work and contributions @oke-aditya !
